Question: Are AndroidLifecycleScopeProvider instances intended to be reusable?

This is for AutoDispose 0.3.0.

A pattern I used for my custom scope providers for fragments/activities against AutoDispose 0.2.0 was to create a single scope provider (based on `BehaviorSubject`) and then share it between all observable chains for that entity. This worked fine, I think.

Now I'm migrating to the new `AndroidLifecycleScopeProvider` and I find that if I create a single instance for a fragment or activity, when, for example, a second `onStart` event occurs I get the exception below.

If I recreate the AndroidLifecycleScopeProvider instance for each subscription then it works fine, but it feels like it could be excess overhead. I do notice that this is the pattern you use in the samples, so perhaps it's intentional. There's no indication in AndroidLifecycleScopeProvider that it's stateful, so I'm just checking.

```
io.reactivex.exceptions.OnErrorNotImplementedException: Lifecycle has ended!
                         E      at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:704)
                         E      at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:701)
                         E      at io.reactivex.internal.observers.LambdaObserver.onError(LambdaObserver.java:77)
                         E      at com.uber.autodispose.AutoDisposingObserverImpl.onError(AutoDisposingObserverImpl.java:102)
                         E      at com.uber.autodispose.AutoDisposingObserverImpl$2.accept(AutoDisposingObserverImpl.java:57)
                         E      at com.uber.autodispose.AutoDisposingObserverImpl$2.accept(AutoDisposingObserverImpl.java:55)
                         E      at io.reactivex.internal.operators.maybe.MaybeCallbackObserver.onError(MaybeCallbackObserver.java:83)
                         E      at io.reactivex.internal.operators.maybe.MaybeDoOnEvent$DoOnEventMaybeObserver.onError(MaybeDoOnEvent.java:100)
                         E      at io.reactivex.internal.disposables.EmptyDisposable.error(EmptyDisposable.java:83)
                         E      at io.reactivex.internal.operators.maybe.MaybeDefer.subscribeActual(MaybeDefer.java:44)
                         E      at io.reactivex.Maybe.subscribe(Maybe.java:3727)
                         E      at io.reactivex.internal.operators.maybe.MaybeDoOnEvent.subscribeActual(MaybeDoOnEvent.java:39)
                         E      at io.reactivex.Maybe.subscribe(Maybe.java:3727)
                         E      at io.reactivex.Maybe.subscribeWith(Maybe.java:3793)
                         E      at io.reactivex.Maybe.subscribe(Maybe.java:3714)
                         E      at io.reactivex.Maybe.subscribe(Maybe.java:3680)
                         E      at com.uber.autodispose.AutoDisposingObserverImpl.onSubscribe(AutoDisposingObserverImpl.java:51)
                         E      at io.reactivex.internal.observers.BasicFuseableObserver.onSubscribe(BasicFuseableObserver.java:66)
                         E      at io.reactivex.internal.observers.BasicFuseableObserver.onSubscribe(BasicFuseableObserver.java:66)
                         E      at com.jakewharton.rxrelay2.PublishRelay.subscribeActual(PublishRelay.java:68)
                         E      at io.reactivex.Observable.subscribe(Observable.java:10910)
                         E      at io.reactivex.internal.operators.observable.ObservableFilter.subscribeActual(ObservableFilter.java:30)
                         E      at io.reactivex.Observable.subscribe(Observable.java:10910)
                         E      at io.reactivex.internal.operators.observable.ObservableMap.subscribeActual(ObservableMap.java:33)
                         E      at io.reactivex.Observable.subscribe(Observable.java:10910)
                         E      at com.uber.autodispose.ObservableScoper$AutoDisposeObservable.subscribeActual(ObservableScoper.java:113)
                         E      at io.reactivex.Observable.subscribe(Observable.java:10910)
                         E      at io.reactivex.Observable.subscribe(Observable.java:10896)
                         E      at io.reactivex.Observable.subscribe(Observable.java:10799)
                         E      at com.uber.autodispose.ObservableScoper$1.subscribe(ObservableScoper.java:71)
                         E      at com.orangebikelabs.BrowseFragment.onStart(BrowseFragment.java:222)
                         E      at android.support.v4.app.Fragment.performStart(Fragment.java:2380)
                         E      at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1458)
                         E      at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1740)
                         E      at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1809)
                         E      at android.support.v4.app.BackStackRecord.executePopOps(BackStackRecord.java:857)
                         E      at android.support.v4.app.FragmentManagerImpl.executeOps(FragmentManager.java:2577)
                         E      at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2367)
                         E      at android.support.v4.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2322)
                         E      at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:2229)
                         E      at android.support.v4.app.FragmentManagerImpl$1.run(FragmentManager.java:700)
                         E      at android.os.Handler.handleCallback(Handler.java:751)
                         E      at android.os.Handler.dispatchMessage(Handler.java:95)
                         E      at android.os.Looper.loop(Looper.java:154)
                         E      at android.app.ActivityThread.main(ActivityThread.java:6120)
                         E      at java.lang.reflect.Method.invoke(Native Method)
                         E      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
                         E      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
                         E  Caused by: com.uber.autodispose.LifecycleEndedException: Lifecycle has ended!
                         E      at com.uber.autodispose.android.lifecycle.AndroidLifecycleScopeProvider$1.apply(AndroidLifecycleScopeProvider.java:52)
                         E      at com.uber.autodispose.android.lifecycle.AndroidLifecycleScopeProvider$1.apply(AndroidLifecycleScopeProvider.java:38)
                         E      at com.uber.autodispose.ScopeUtil$3.call(ScopeUtil.java:91)
                         E      at com.uber.autodispose.ScopeUtil$3.call(ScopeUtil.java:74)
                         E      at io.reactivex.internal.operators.maybe.MaybeDefer.subscribeActual(MaybeDefer.java:41)
                         E      ... 39 more
```

--- a/android/autodispose-android-archcomponents/src/main/java/com/uber/autodispose/android/lifecycle/AndroidLifecycleScopeProvider.java
+++ b/android/autodispose-android-archcomponents/src/main/java/com/uber/autodispose/android/lifecycle/AndroidLifecycleScopeProvider.java
@@ -49,7 +49,7 @@ public final class AndroidLifecycleScopeProvider
             case ON_STOP:
             case ON_DESTROY:
             default:
-              throw new LifecycleEndedException();
+              throw new LifecycleEndedException("Lifecycle has ended! Last event was " + lastEvent);
           }
         }
       };
@@ -145,6 +145,7 @@ public final class AndroidLifecycleScopeProvider
   }
 
   @Override public Lifecycle.Event peekLifecycle() {
+    lifecycleObservable.backfillEvents();
     return lifecycleObservable.getValue();
   }
 

--- a/android/autodispose-android-archcomponents/src/main/java/com/uber/autodispose/android/lifecycle/LifecycleEventsObservable.java
+++ b/android/autodispose-android-archcomponents/src/main/java/com/uber/autodispose/android/lifecycle/LifecycleEventsObservable.java
@@ -48,10 +48,12 @@ import static com.uber.autodispose.android.internal.AutoDisposeAndroidUtil.isMai
     return eventsObservable.getValue();
   }
 
+  /**
+   * Backfill if already created for boundary checking. We do a trick here for corresponding events
+   * where we pretend something is created upon initialized state so that it assumes the
+   * corresponding event is DESTROY.
+   */
   void backfillEvents() {
-    // Backfill if already created for boundary checking
-    // We do a trick here for corresponding events where we pretend something is created
-    // upon initialized state so that it assumes the corresponding event is DESTROY.
     @Nullable Lifecycle.Event correspondingEvent;
     switch (lifecycle.getCurrentState()) {
       case INITIALIZED:

--- a/android/autodispose-android-archcomponents/src/main/java/com/uber/autodispose/android/lifecycle/LifecycleEventsObservable.java
+++ b/android/autodispose-android-archcomponents/src/main/java/com/uber/autodispose/android/lifecycle/LifecycleEventsObservable.java
@@ -38,31 +38,34 @@ import static com.uber.autodispose.android.internal.AutoDisposeAndroidUtil.isMai
 
   @SuppressWarnings("CheckReturnValue") LifecycleEventsObservable(Lifecycle lifecycle) {
     this.lifecycle = lifecycle;
-    // Backfill if already created for boundary checking
-    // We do a trick here for corresponding events where we pretend something is created
-    // upon initialized state so that it assumes the corresponding event is DESTROY.
-    @Nullable Event correspondingEvent;
-    switch (lifecycle.getCurrentState()) {
-      case INITIALIZED:
-        correspondingEvent = Event.ON_CREATE;
-        break;
-      case CREATED:
-        correspondingEvent = Event.ON_START;
-        break;
-      case STARTED:
-      case RESUMED:
-        correspondingEvent = Event.ON_RESUME;
-        break;
-      case DESTROYED:
-      default:
-        correspondingEvent = Event.ON_DESTROY;
-        break;
-    }
-    eventsObservable.onNext(correspondingEvent);
   }
 
   Event getValue() {
     return eventsObservable.getValue();
+  }
+
+  void backfillEvents() {
+    // Backfill if already created for boundary checking
+    // We do a trick here for corresponding events where we pretend something is created
+    // upon initialized state so that it assumes the corresponding event is DESTROY.
+    @Nullable Lifecycle.Event correspondingEvent;
+    switch (lifecycle.getCurrentState()) {
+      case INITIALIZED:
+        correspondingEvent = Lifecycle.Event.ON_CREATE;
+        break;
+      case CREATED:
+        correspondingEvent = Lifecycle.Event.ON_START;
+        break;
+      case STARTED:
+      case RESUMED:
+        correspondingEvent = Lifecycle.Event.ON_RESUME;
+        break;
+      case DESTROYED:
+      default:
+        correspondingEvent = Lifecycle.Event.ON_DESTROY;
+        break;
+    }
+    eventsObservable.onNext(correspondingEvent);
   }
 
   @Override protected void subscribeActual(Observer<? super Event> observer) {
@@ -97,7 +100,7 @@ import static com.uber.autodispose.android.internal.AutoDisposeAndroidUtil.isMai
     @OnLifecycleEvent(Event.ON_ANY) void onStateChange(LifecycleOwner owner, Event event) {
       if (!isDisposed()) {
         if (!(event == Event.ON_CREATE && eventsObservable.getValue() == event)) {
-          // Due to the INITIALIZED->ON_CREATE mapping trick we do in the constructor backfill,
+          // Due to the INITIALIZED->ON_CREATE mapping trick we do in backfill(),
           // we fire this conditionally to avoid duplicate CREATE events.
           eventsObservable.onNext(event);
         }

--- a/android/autodispose-android-archcomponents/src/main/java/com/uber/autodispose/android/lifecycle/LifecycleEventsObservable.java
+++ b/android/autodispose-android-archcomponents/src/main/java/com/uber/autodispose/android/lifecycle/LifecycleEventsObservable.java
@@ -28,6 +28,10 @@ import io.reactivex.Observer;
 import io.reactivex.android.MainThreadDisposable;
 import io.reactivex.subjects.BehaviorSubject;
 
+import static android.arch.lifecycle.Lifecycle.Event.ON_CREATE;
+import static android.arch.lifecycle.Lifecycle.Event.ON_DESTROY;
+import static android.arch.lifecycle.Lifecycle.Event.ON_RESUME;
+import static android.arch.lifecycle.Lifecycle.Event.ON_START;
 import static android.support.annotation.RestrictTo.Scope.LIBRARY;
 import static com.uber.autodispose.android.internal.AutoDisposeAndroidUtil.isMainThread;
 
@@ -51,18 +55,18 @@ import static com.uber.autodispose.android.internal.AutoDisposeAndroidUtil.isMai
     @Nullable Lifecycle.Event correspondingEvent;
     switch (lifecycle.getCurrentState()) {
       case INITIALIZED:
-        correspondingEvent = Lifecycle.Event.ON_CREATE;
+        correspondingEvent = ON_CREATE;
         break;
       case CREATED:
-        correspondingEvent = Lifecycle.Event.ON_START;
+        correspondingEvent = ON_START;
         break;
       case STARTED:
       case RESUMED:
-        correspondingEvent = Lifecycle.Event.ON_RESUME;
+        correspondingEvent = ON_RESUME;
         break;
       case DESTROYED:
       default:
-        correspondingEvent = Lifecycle.Event.ON_DESTROY;
+        correspondingEvent = ON_DESTROY;
         break;
     }
     eventsObservable.onNext(correspondingEvent);
@@ -99,7 +103,7 @@ import static com.uber.autodispose.android.internal.AutoDisposeAndroidUtil.isMai
 
     @OnLifecycleEvent(Event.ON_ANY) void onStateChange(LifecycleOwner owner, Event event) {
       if (!isDisposed()) {
-        if (!(event == Event.ON_CREATE && eventsObservable.getValue() == event)) {
+        if (!(event == ON_CREATE && eventsObservable.getValue() == event)) {
           // Due to the INITIALIZED->ON_CREATE mapping trick we do in backfill(),
           // we fire this conditionally to avoid duplicate CREATE events.
           eventsObservable.onNext(event);

--- a/sample/src/main/kotlin/com/uber/autodispose/sample/KotlinActivity.kt
+++ b/sample/src/main/kotlin/com/uber/autodispose/sample/KotlinActivity.kt
@@ -32,6 +32,9 @@ import java.util.concurrent.TimeUnit
  */
 class KotlinActivity : AppCompatActivity() {
 
+  // Can be reused
+  private val scopeProvider by lazy { AndroidLifecycleScopeProvider.from(this) }
+
   override fun onCreate(savedInstanceState: Bundle?) {
     super.onCreate(savedInstanceState)
     Log.d(TAG, "onCreate()")
@@ -41,7 +44,7 @@ class KotlinActivity : AppCompatActivity() {
     // dispose is onDestroy (the opposite of onCreate).
     Observable.interval(1, TimeUnit.SECONDS)
         .doOnDispose { Log.i(TAG, "Disposing subscription from onCreate()") }
-        .autoDisposeWith(AndroidLifecycleScopeProvider.from(this))
+        .autoDisposeWith(scopeProvider)
         .subscribe { num -> Log.i(TAG, "Started in onCreate(), running until onDestroy(): " + num) }
   }
 
@@ -54,7 +57,7 @@ class KotlinActivity : AppCompatActivity() {
     // dispose is onStop (the opposite of onStart).
     Observable.interval(1, TimeUnit.SECONDS)
         .doOnDispose { Log.i(TAG, "Disposing subscription from onStart()") }
-        .autoDisposeWith(AndroidLifecycleScopeProvider.from(this))
+        .autoDisposeWith(scopeProvider)
         .subscribe { num -> Log.i(TAG, "Started in onStart(), running until in onStop(): " + num) }
   }
 
@@ -67,7 +70,7 @@ class KotlinActivity : AppCompatActivity() {
     // dispose is onPause (the opposite of onResume).
     Observable.interval(1, TimeUnit.SECONDS)
         .doOnDispose { Log.i(TAG, "Disposing subscription from onResume()") }
-        .autoDisposeWith(AndroidLifecycleScopeProvider.from(this))
+        .autoDisposeWith(scopeProvider)
         .subscribe { num ->
           Log.i(TAG, "Started in onResume(), running until in onPause(): " + num)
         }

